src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStep2Test.java

```diff
@@ -185,4 +189,35 @@ private TestCloudSlave(String name) throws Exception {
         }
     }
 
+    @Test public void selfNameVsLabel() throws Throwable {
+        logging.recordPackage(ExecutorStepExecution.class, Level.FINE);
+        rr.then(r -> {
+            r.jenkins.setNumExecutors(0);
+            ExtensionList.lookupSingleton(DelayX.class).active = false;
+            r.createSlave("x", null, null);
+            r.createSlave("x2", "x", null);
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition("parallel b1: {node('x && !x2') {semaphore 'b1'}}, b2: {node('x2') {semaphore 'b2'}}", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("b1/1", b);
+            SemaphoreStep.waitForStart("b2/1", b);
+        });
+        rr.then(r -> {
+            WorkflowJob p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+            WorkflowRun b = p.getBuildByNumber(1);
+            SemaphoreStep.success("b1/1", null);
+            SemaphoreStep.success("b2/1", null);
+            r.waitForCompletion(b);
+            r.assertBuildStatusSuccess(b);
+        });
+    }
+    @TestExtension("selfNameVsLabel") public static final class DelayX extends ComputerListener {
+        boolean active = true;
+        @Override public void preOnline(Computer c, Channel channel, FilePath root, TaskListener listener) throws IOException, InterruptedException {
+            if (c.getName().equals("x")) {
+                Thread.sleep(5_000);
```

Comment:
Ditto; trying to make `x2` come back up first so that the task originally run on `x` tries to get scheduled on it.